From 3bbf2dbe254139352f6ad9c2c5ea3e03a05e943d Mon Sep 17 00:00:00 2001
From: Ranjan Dutta <ranjan.dutta@intel.com>
Date: Thu, 27 Apr 2023 13:18:21 +0800
Subject: [PATCH 65/68] Revert "drm/i915/gvt: fix double free bug in
 split_2MB_gtt_entry"

This reverts commit 3d743415c6fb092167df6c23e9c7e9f6df7db625.
---
 drivers/gpu/drm/i915/gvt/gtt.c | 17 ++++-------------
 1 file changed, 4 insertions(+), 13 deletions(-)

diff --git a/drivers/gpu/drm/i915/gvt/gtt.c b/drivers/gpu/drm/i915/gvt/gtt.c
index 0201f9b5f87e7..a3a4305eda01b 100644
--- a/drivers/gpu/drm/i915/gvt/gtt.c
+++ b/drivers/gpu/drm/i915/gvt/gtt.c
@@ -1192,8 +1192,10 @@ static int split_2MB_gtt_entry(struct intel_vgpu *vgpu,
 	for_each_shadow_entry(sub_spt, &sub_se, sub_index) {
 		ret = intel_gvt_hypervisor_dma_map_guest_page(vgpu,
 				start_gfn + sub_index, PAGE_SIZE, &dma_addr);
-		if (ret)
-			goto err;
+		if (ret) {
+			ppgtt_invalidate_spt(spt);
+			return ret;
+		}
 		sub_se.val64 = se->val64;
 
 		/* Copy the PAT field from PDE. */
@@ -1212,17 +1214,6 @@ static int split_2MB_gtt_entry(struct intel_vgpu *vgpu,
 	ops->set_pfn(se, sub_spt->shadow_page.mfn);
 	ppgtt_set_shadow_entry(spt, se, index);
 	return 0;
-err:
-	/* Cancel the existing addess mappings of DMA addr. */
-	for_each_present_shadow_entry(sub_spt, &sub_se, sub_index) {
-		gvt_vdbg_mm("invalidate 4K entry\n");
-		ppgtt_invalidate_pte(sub_spt, &sub_se);
-	}
-	/* Release the new allocated spt. */
-	trace_spt_change(sub_spt->vgpu->id, "release", sub_spt,
-		sub_spt->guest_page.gfn, sub_spt->shadow_page.type);
-	ppgtt_free_spt(sub_spt);
-	return ret;
 }
 
 static int split_64KB_gtt_entry(struct intel_vgpu *vgpu,
-- 
2.25.1

